File: src/main/java/com/hileco/cortex/analysis/Graph.java
package com.hileco.cortex.analysis;

import com.hileco.cortex.analysis.edges.Edge;
import com.hileco.cortex.instructions.Instruction;
import com.hileco.cortex.instructions.jumps.JUMP_DESTINATION;
import lombok.Getter;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicReference;
import java.util.stream.Collectors;

public class Graph {
    @Getter
    private final List<GraphBlock> graphBlocks;
    @Getter
    private final List<Edge> edges;

    public Graph() {
        this.graphBlocks = new ArrayList<>();
        this.edges = new ArrayList<>();
    }

    public Graph(List<Instruction> instructions) {
        this();
        var block = new ArrayList<AtomicReference<Instruction>>();
        var currentBlockLine = 0;
        var currentLine = 0;
        while (currentLine < instructions.size()) {
            AtomicReference<Instruction> instructionReference = new AtomicReference<>(instructions.get(currentLine));
            if (instructionReference.get() instanceof JUMP_DESTINATION) {
                if (!block.isEmpty()) {
                    this.includeAsBlock(currentBlockLine, block);
                    currentBlockLine = currentLine;
                }
                block.clear();
            }
            block.add(instructionReference);
            currentLine++;
        }
        if (!block.isEmpty()) {
            this.includeAsBlock(currentBlockLine, block);
        }
    }

    private void includeAsBlock(int line, List<AtomicReference<Instruction>> instructions) {
        var block = new GraphBlock();
        block.include(line, instructions);
        this.graphBlocks.add(block);
    }

    private int indexOf(GraphBlock graphBlock) {
        var index = this.graphBlocks.indexOf(graphBlock);
        if (index == -1) {
            throw new IllegalArgumentException();
        }
        return index;
    }

    public void mergeUpwards(GraphBlock graphBlock) {
        var index = this.indexOf(graphBlock);
        if (index == 0) {
            return;
        }
        var target = this.graphBlocks.get(index - 1);
        target.append(graphBlock);
        this.graphBlocks.remove(graphBlock);
    }

    public void mergeDownwards(GraphBlock graphBlock) {
        var index = this.indexOf(graphBlock);
        if (index + 1 >= this.graphBlocks.size()) {
            return;
        }
        var target = this.graphBlocks.get(index + 1);
        graphBlock.append(target);
        this.graphBlocks.remove(target);
    }

    public void remove(GraphBlock graphBlock) {
        this.graphBlocks.remove(graphBlock);
    }

    public void replace(GraphBlock original, GraphBlock replacement) {
        var index = this.indexOf(original);
        this.graphBlocks.set(index, replacement);
    }

    public List<Instruction> toInstructions() {
        return this.graphBlocks.stream()
                .flatMap(graphBlock -> graphBlock.getGraphNodes().stream())
                .map(graphNode -> graphNode.getInstruction().get())
                .collect(Collectors.toList());
    }

    @Override
    public String toString() {
        var stringBuilder = new StringBuilder();
        for (var graphBlock : this.graphBlocks) {
            stringBuilder.append(graphBlock);
            stringBuilder.append("\n");
        }
        return stringBuilder.toString();
    }
}

File: src/main/java/com/hileco/cortex/analysis/Graph.kt
package com.hileco.cortex.analysis

import com.hileco.cortex.analysis.edges.EdgeMapping
import com.hileco.cortex.instructions.Instruction
import com.hileco.cortex.instructions.jumps.JUMP_DESTINATION
import java.util.*

class Graph() {
    val graphBlocks: MutableList<GraphBlock> = ArrayList()
    val edgeMapping: EdgeMapping = EdgeMapping()

    constructor(instructions: List<Instruction>) : this() {
        val block = ArrayList<Instruction>()
        var currentBlockLine = 0
        for (currentLine in 0 until instructions.size) {
            val instruction = instructions[currentLine]
            if (instruction is JUMP_DESTINATION) {
                if (!block.isEmpty()) {
                    includeAsBlock(currentBlockLine, block)
                    currentBlockLine = currentLine
                }
                block.clear()
            }
            block.add(instruction)
        }
        if (!block.isEmpty()) {
            includeAsBlock(currentBlockLine, block)
        }
    }

    private fun includeAsBlock(line: Int, instructions: List<Instruction>) {
        val block = GraphBlock()
        block.include(line, instructions)
        graphBlocks.add(block)
    }

    private fun indexOf(graphBlock: GraphBlock): Int {
        val index = graphBlocks.indexOf(graphBlock)
        if (index == -1) {
            throw IllegalArgumentException()
        }
        return index
    }

    fun blockAt(line: Int): GraphBlock? {
        return graphBlocks.firstOrNull { graphBlock: GraphBlock ->
            graphBlock.graphNodes.any { graphNode: GraphNode ->
                graphNode.line == line
            }
        }
    }

    fun mergeUpwards(graphBlock: GraphBlock) {
        val index = this.indexOf(graphBlock)
        if (index == 0) {
            return
        }
        val target = graphBlocks[index - 1]
        target.append(graphBlock)
        graphBlocks.remove(graphBlock)
    }

    fun mergeDownwards(graphBlock: GraphBlock) {
        val index = this.indexOf(graphBlock)
        if (index + 1 >= graphBlocks.size) {
            return
        }
        val target = graphBlocks[index + 1]
        graphBlock.append(target)
        graphBlocks.remove(target)
    }

    fun remove(graphBlock: GraphBlock) {
        graphBlocks.remove(graphBlock)
    }

    fun replace(original: GraphBlock, replacement: GraphBlock) {
        val index = this.indexOf(original)
        graphBlocks[index] = replacement
    }

    fun toInstructions(): List<Instruction> {
        return graphBlocks.asSequence()
                .flatMap { it.graphNodes.asSequence() }
                .map { it.instruction }
                .toList()
    }

    override fun toString(): String {
        return graphBlocks.joinToString { "$it\n" }
    }
}
